pkg/spdk/replica.go

```diff
@@ -1543,6 +1609,558 @@ func (r *Replica) SnapshotRangeHashGet(spdkClient *spdkclient.Client, snapshotNa
+	updateRequired := false
+
+	r.Lock()
+	defer func() {
+		r.Unlock()
+
+		if updateRequired {
+			r.UpdateCh <- nil
+		}
+	}()
+
+	if r.isSnapshotCloning {
+		return fmt.Errorf("replica %s cloning is in process", r.Name)
+	}
+	r.isSnapshotCloning = true
+
+	defer func() {
+		if err != nil {
+			if r.State != types.InstanceStateError {
+				r.State = types.InstanceStateError
+			}
+			r.ErrorMsg = err.Error()
+			if r.snapshotCloningDstCache.cloningError == "" {
+				r.snapshotCloningDstCache.cloningError = err.Error()
+				r.snapshotCloningDstCache.cloningState = types.ProgressStateError
+			}
+		} else {
+			if r.State != types.InstanceStateError {
+				r.ErrorMsg = ""
+			}
+		}
+
+		updateRequired = true
+	}()
+
+	// Replica.Delete and Replica.Create do not guarantee that the previous cloning dst replica info is cleaned up
+	if err := r.doCleanupForSnapshotCloneDst(spdkClient, true); err != nil {
+		return errors.Wrapf(err, "failed to clean up the previous cloning dst info for dst replica snapshot "+
+			"clone start, src replica name %s, address %s, snapshot name %s", r.snapshotCloningDstCache.srcReplicaName,
+			r.snapshotCloningDstCache.srcReplicaAddress, r.snapshotCloningDstCache.snapshotName)
+	}
+	// init cloning
+	r.snapshotCloningDstCache.snapshotName = snapshotName
+	r.snapshotCloningDstCache.srcReplicaName = srcReplicaName
+	r.snapshotCloningDstCache.srcReplicaAddress = srcReplicaAddress
+
+	if cloneMode == spdkrpc.CloneMode_CLONE_MODE_LINKED_CLONE {
+		srcReplicaIP, _, err := splitHostPort(srcReplicaAddress)
+		if err != nil {
+			return errors.Wrapf(err, "failed to split src Replica address %v", srcReplicaAddress)
+		}
+
+		if r.IP != srcReplicaIP {
+			return fmt.Errorf("failed to do snapshot linked-clone: dst replica IP %v is not the same as "+
+				"src replica IP %v", r.IP, srcReplicaIP)
+		}
+
+		srcReplicaServiceCli, err := GetServiceClient(r.snapshotCloningDstCache.srcReplicaAddress)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if errClose := srcReplicaServiceCli.Close(); errClose != nil {
+				r.log.WithError(errClose).Errorf("Failed to close replica %s client with address %s during "+
+					"start cloning at dst", r.snapshotCloningDstCache.srcReplicaName, r.snapshotCloningDstCache.srcReplicaAddress)
+			}
+		}()
+
+		if err := srcReplicaServiceCli.ReplicaSnapshotCloneSrcStart(r.snapshotCloningDstCache.srcReplicaName,
+			snapshotName, r.Name, "", cloneMode); err != nil {
+			return err
+		}
+		r.snapshotCloningDstCache.cloningState = types.ProgressStateComplete
+		r.log.Infof("Clone state: %v", r.snapshotCloningDstCache.cloningState)
+		return r.SnapshotCloneDstFinish(spdkClient, cloneMode)
+	}
+
+	if r.snapshotCloningDstCache.cloningPort == 0 {
+		if r.snapshotCloningDstCache.cloningPort, _, err = r.portAllocator.AllocateRange(1); err != nil {
+			return errors.Wrapf(err, "failed to allocate a cloning port for dst replica %v snapshot clone start", r.Name)
+		}
```

Comment:
Do we need a defer func to release the port if there is any error?

Reply:
No, it will be released in the monitor `monitorSnapshotClone`